azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,577 @@
+        if requested_version < CURRENT_VERSION:
+            # In case of a downgrade, we mark the current agent as bad version to avoid starting it back up ever again
+            # (the expectation here being that if we get request to a downgrade,
+            # there's a good reason for not wanting the current version).
+            prefix = "downgrade"
+            try:
+                # We should always have an agent directory for the CURRENT_VERSION
+                # (unless the CURRENT_VERSION == daemon version, but since we don't support downgrading
+                # below daemon version, we will never reach this code path if that's the scenario)
+                agents_on_disk = AgentUpdateHandler.__get_available_agents_on_disk()
+                current_agent = next(agent for agent in agents_on_disk if agent.version == CURRENT_VERSION)
+                msg = "Marking the agent {0} as bad version since a downgrade was requested in the GoalState, " \
+                      "suggesting that we really don't want to execute any extensions using this version".format(CURRENT_VERSION)
+                logger.info(msg)
+                current_agent.mark_failure(is_fatal=True, reason=msg)
+            except StopIteration:
+                logger.warn(
+                    "Could not find a matching agent with current version {0} to blacklist, skipping it".format(
+                        CURRENT_VERSION))
+        else:
+            # In case of an upgrade, we don't need to exclude anything as the daemon will automatically
+            # start the next available highest version which would be the target version
+            prefix = "upgrade"
+        raise AgentUpgradeExitException("Agent update found, Exiting current process to {0} to the new Agent version {1}".format(prefix, requested_version))
+
+    @staticmethod
+    def __get_available_agents_on_disk():
+        available_agents = [agent for agent in AgentUpdateHandler.__get_all_agents_on_disk() if not agent.is_blacklisted]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    @staticmethod
+    def __get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if os.path.isdir(agent_dir)]
+
+    @staticmethod
+    def __log_event(level, msg_, success_=True):
+        if level == LogLevel.WARNING:
+            logger.warn(msg_)
+        elif level == LogLevel.ERROR:
+            logger.error(msg_)
+        elif level == LogLevel.INFO:
+            logger.info(msg_)
+        add_event(op=WALAEventOperation.AgentUpgrade, is_success=success_, message=msg_, log_event=False)
+
+    def run(self):
+        try:
+            # Ignore new agents if update is disabled
+            if not self.autoupdate_enabled:
+                return
+
+            agent_manifest = self.__get_agent_manifest_from_last_gs()
```

Comment:
Expressing this in terms of manifest makes the code confusing, since we do not need the agent manifests to get the required version, but just the extensions goal state.

Reply:
ok, I'll rename